File: ios/Bristlecone/Sources/Storage.swift
import Foundation

/// Settings, saved places and field reports on the device, plus backup and restore. The layout
/// under Application Support matches the Android app's files folder (kv/, tracks/, regions/,
/// cache/), and so does the backup zip.
final class Storage {
    let root: URL
    let kvDir: URL
    private let queue = DispatchQueue(label: "bristlecone.storage")

    init(root: URL) {
        self.root = root
        kvDir = root.appendingPathComponent("kv", isDirectory: true)
        try? FileManager.default.createDirectory(at: kvDir, withIntermediateDirectories: true)
    }

    static func safe(_ key: String) -> String {
        String(key.map { c -> Character in
            c.isASCII && (c.isLetter || c.isNumber || c == "_" || c == "." || c == "-") ? c : "_"
        })
    }

    private func file(_ key: String) -> URL { kvDir.appendingPathComponent(Storage.safe(key) + ".json") }

    /// Every stored value, handed to the page at startup so reads never wait on the bridge.
    func all() -> [String: String] {
        queue.sync {
            var out: [String: String] = [:]
            let names = (try? FileManager.default.contentsOfDirectory(atPath: kvDir.path)) ?? []
            for n in names where n.hasSuffix(".json") {
                if let s = try? String(contentsOf: kvDir.appendingPathComponent(n), encoding: .utf8) {
                    out[String(n.dropLast(5))] = s
                }
            }
            return out
        }
    }

    func put(_ key: String, _ value: String) {
        let f = file(key)
        queue.async { try? Data(value.utf8).write(to: f, options: .atomic) }
    }

    func remove(_ key: String) {
        let f = file(key)
        queue.async { try? FileManager.default.removeItem(at: f) }
    }

    func kvBytes() -> Int64 { Storage.folderBytes(kvDir) }

    static func folderBytes(_ dir: URL) -> Int64 {
        let fm = FileManager.default
        let names = (try? fm.contentsOfDirectory(atPath: dir.path)) ?? []
        return names.reduce(Int64(0)) { sum, n in
            let a = try? fm.attributesOfItem(atPath: dir.appendingPathComponent(n).path)
            return sum + ((a?[.size] as? NSNumber)?.int64Value ?? 0)
        }
    }

    // MARK: Backup and restore

    /// Writes kv/ and tracks/ (and optionally regions/ and cache/) to a zip. Returns the file count.
    func backup(to url: URL, includeMaps: Bool) throws -> Int {
        try queue.sync {
            let w = try Zip.Writer(url: url)
            let created = Int64(Date().timeIntervalSince1970 * 1000)
            try w.add(name: "bristlecone-backup.txt", data: Data("Bristlecone backup\nformat=1\ncreated=\(created)\nmaps=\(includeMaps)\n".utf8))
            var n = 0
            var folders = ["kv", "tracks"]
            if includeMaps { folders += ["regions", "cache"] }
            let fm = FileManager.default
            for folder in folders {
                let dir = root.appendingPathComponent(folder, isDirectory: true)
                for name in ((try? fm.contentsOfDirectory(atPath: dir.path)) ?? []).sorted() where !name.hasSuffix(".tmp") {
                    var isDir: ObjCBool = false
                    let f = dir.appendingPathComponent(name)
                    guard fm.fileExists(atPath: f.path, isDirectory: &isDir), !isDir.boolValue,
                          let d = try? Data(contentsOf: f) else { continue }
                    try w.add(name: folder + "/" + name, data: d)
                    n += 1
                }
            }
            try w.finish()
            return n
        }
    }

    /// Restores a backup zip. Only the known folders are written; anything else is ignored.
    func restore(from url: URL) throws -> Int {
        try queue.sync {
            let z = try Data(contentsOf: url, options: .mappedIfSafe)
            let entries = try Zip.entries(z)
            guard entries.contains(where: { $0.name == "bristlecone-backup.txt" }) else {
                throw Zip.Failure(message: "not a Bristlecone backup")
            }
            var n = 0
            let fm = FileManager.default
            for e in entries {
                let parts = e.name.split(separator: "/", omittingEmptySubsequences: false).map(String.init)
                guard parts.count == 2, ["kv", "tracks", "regions", "cache"].contains(parts[0]),
                      !parts[1].isEmpty, !parts[1].contains(".."), !parts[1].hasPrefix(".") else { continue }
                let dir = root.appendingPathComponent(parts[0], isDirectory: true)
                try? fm.createDirectory(at: dir, withIntermediateDirectories: true)
                try Zip.read(z, e).write(to: dir.appendingPathComponent(parts[1]), options: .atomic)
                n += 1
            }
            return n
        }
    }
}

File: app/src/main/java/app/bristlecone/Storage.java
package app.bristlecone;

import android.content.Context;

import org.json.JSONArray;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;
import java.util.zip.ZipOutputStream;

/**
 * On-phone storage for settings, saved places and field reports, plus backup and restore.
 * Backups are a single .zip written through Android's document picker, so the user can
 * send them straight to Google Drive (or anywhere else) with no account setup in the app.
 */
public final class Storage {
    private final File kvDir;
    private final File root;

    public Storage(Context ctx) {
        root = ctx.getFilesDir();
        kvDir = new File(root, "kv");
        //noinspection ResultOfMethodCallIgnored
        kvDir.mkdirs();
    }

    static String safe(String key) { return key.replaceAll("[^A-Za-z0-9_.-]", "_"); }

    public synchronized String get(String key) {
        File f = new File(kvDir, safe(key) + ".json");
        if (!f.exists()) return null;
        try { return new String(NetCache.readAll(new FileInputStream(f)), "UTF-8"); } catch (IOException e) { return null; }
    }

    public synchronized boolean put(String key, String value) {
        try {
            File tmp = new File(kvDir, safe(key) + ".json.tmp");
            FileOutputStream o = new FileOutputStream(tmp);
            o.write(value.getBytes("UTF-8"));
            o.getFD().sync();
            o.close();
            return tmp.renameTo(new File(kvDir, safe(key) + ".json"));
        } catch (IOException e) {
            return false;
        }
    }

    public synchronized void remove(String key) {
        //noinspection ResultOfMethodCallIgnored
        new File(kvDir, safe(key) + ".json").delete();
    }

    public synchronized JSONArray keys(String prefix) {
        JSONArray a = new JSONArray();
        File[] fs = kvDir.listFiles();
        if (fs != null) for (File f : fs) {
            String n = f.getName();
            if (!n.endsWith(".json")) continue;
            n = n.substring(0, n.length() - 5);
            if (prefix == null || n.startsWith(safe(prefix))) a.put(n);
        }
        return a;
    }

    public long kvBytes() {
        long t = 0;
        File[] fs = kvDir.listFiles();
        if (fs != null) for (File f : fs) t += f.length();
        return t;
    }

    // ------------------------------------------------------------------ Backup and restore

    /** Writes kv/ (and optionally regions/ and cache/) to a zip. Returns number of files. */
    public int backup(OutputStream out, boolean includeMaps) throws IOException {
        ZipOutputStream zo = new ZipOutputStream(out);
        int n = 0;
        try {
            ZipEntry marker = new ZipEntry("bristlecone-backup.txt");
            zo.putNextEntry(marker);
            zo.write(("Bristlecone backup\nformat=1\ncreated=" + System.currentTimeMillis() + "\nmaps=" + includeMaps + "\n").getBytes("UTF-8"));
            zo.closeEntry();
            n += addDir(zo, kvDir, "kv/");
            n += addDir(zo, new File(root, "tracks"), "tracks/");
            if (includeMaps) {
                n += addDir(zo, new File(root, "regions"), "regions/");
                n += addDir(zo, new File(root, "cache"), "cache/");
            }
        } finally {
            zo.finish();
            zo.flush();
            out.close();
        }
        return n;
    }

    private int addDir(ZipOutputStream zo, File dir, String prefix) throws IOException {
        File[] fs = dir.listFiles();
        if (fs == null) return 0;
        int n = 0;
        byte[] buf = new byte[32768];
        for (File f : fs) {
            if (!f.isFile() || f.getName().endsWith(".tmp")) continue;
            zo.putNextEntry(new ZipEntry(prefix + f.getName()));
            FileInputStream in = new FileInputStream(f);
            int r;
            while ((r = in.read(buf)) > 0) zo.write(buf, 0, r);
            in.close();
            zo.closeEntry();
            n++;
        }
        return n;
    }

    /** Restores a backup zip. Only known folders are written; anything else is ignored. */
    public int restore(InputStream in) throws IOException {
        ZipInputStream zi = new ZipInputStream(in);
        int n = 0;
        boolean valid = false;
        byte[] buf = new byte[32768];
        try {
            ZipEntry e;
            while ((e = zi.getNextEntry()) != null) {
                String name = e.getName();
                if (name.equals("bristlecone-backup.txt")) { valid = true; continue; }
                if (!valid) throw new IOException("not a Bristlecone backup");
                if (e.isDirectory() || name.contains("..")) continue;
                String folder;
                if (name.startsWith("kv/")) folder = "kv";
                else if (name.startsWith("tracks/")) folder = "tracks";
                else if (name.startsWith("regions/")) folder = "regions";
                else if (name.startsWith("cache/")) folder = "cache";
                else continue;
                String file = name.substring(folder.length() + 1);
                if (file.contains("/") || file.isEmpty()) continue;
                File dir = new File(root, folder);
                //noinspection ResultOfMethodCallIgnored
                dir.mkdirs();
                FileOutputStream o = new FileOutputStream(new File(dir, file));
                int r;
                while ((r = zi.read(buf)) > 0) o.write(buf, 0, r);
                o.close();
                n++;
            }
        } finally {
            zi.close();
        }
        if (!valid) throw new IOException("not a Bristlecone backup");
        return n;
    }
}
